File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque
def solution(bridge_length, weight, truck_weights):
    answer = 0
    
    w=0 # 다리에 올라간 트럭들 무게
    wq=deque(truck_weights) # 대기트럭 큐
    bq=deque([0 for _ in range(bridge_length)]) # 다리 큐

    time=0
    while len(wq) or w>0:
        w-=bq.popleft()

        if len(wq) and w+wq[0]<=weight:
            n = wq.popleft()
            w+=n
            bq.append(n)
        else:
            bq.append(0)
        
        time+=1   
    return time

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;
class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0; // 소요시간
                 

        
        Queue<Integer> bridge = new LinkedList<>();
        for (int i=0; i<bridge_length; i++){
            bridge.add(0); 
        }
        
         Queue<Integer> waitTrucks = new LinkedList<>();
        for (int i=0; i<truck_weights.length; i++){
            waitTrucks.add(truck_weights[i]); 
        }
        
        
        int bridgeWeight = 0; // 다리 무게

        
        while(waitTrucks.size()>0){
            bridgeWeight -= bridge.poll();
                        
            if (waitTrucks.size()>0 && bridgeWeight + waitTrucks.peek() <= weight){ // 견딜수 있는 무게
                int truckWeight = waitTrucks.poll();
                bridge.add(truckWeight);
                bridgeWeight+=truckWeight;
            }
            else{
                bridge.add(0);
            }
            answer++;
        }
        
        
        while(bridgeWeight>0){
            bridgeWeight-=bridge.poll();
            answer++;
        }
        
        return answer;
    }
}